README.md
# rasalas
android app to confirm prediction results of FX


CHANGELOG.md
# 1.1.0 (2019/10/21)
- [UPDATE] prediction view to index multiple pairs

# 1.0.0 (2019/09/25)
- [NEW] create app


app/src/test/java/com/leonis/android/rasalas/ExampleUnitTest.java
package com.leonis.android.rasalas;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

app/src/androidTest/java/MainActivityTest.java
import android.support.test.InstrumentationRegistry;
import android.support.test.runner.AndroidJUnit4;
import android.test.ActivityInstrumentationTestCase2;

import com.leonis.android.rasalas.MainActivity;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

/**
 * Created by leonis on 2019/09/24.
 */

@RunWith(AndroidJUnit4.class)
public class MainActivityTest extends ActivityInstrumentationTestCase2<MainActivity> {
    public MainActivityTest() {
        super(MainActivity.class);
    }

    @Before
    public void setUp() throws Exception {
        super.setUp();
        injectInstrumentation(InstrumentationRegistry.getInstrumentation());
    }

    @After
    public void tearDown() throws Exception {
        super.tearDown();
    }

    @Test
    public void testOnCreate() {
        assertEquals(true, true);
    }
}


app/src/main/java/com/leonis/android/rasalas/MainActivity.java
package com.leonis.android.rasalas;

import android.app.LoaderManager;
import android.content.Context;
import android.content.Loader;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;

import com.leonis.android.rasalas.lib.HTTPClient;
import com.leonis.android.rasalas.models.Prediction;
import com.leonis.android.rasalas.views.PredictionView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;

public class MainActivity extends AppCompatActivity {
    private static final int LOADER_ID = 0;

    private Context activity;
    private PredictionView predictionView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        activity = this;
        predictionView = findViewById(R.id.prediction);
        getPredictions(PredictionView.DEFAULT_PAGE, PredictionView.DEFAULT_PAIR);
    }

    public void getPredictions(String page, final String pair) {
        Bundle args = new Bundle();
        args.putString("page", page);
        args.putString("pair", pair);

        getLoaderManager().initLoader(LOADER_ID, args, new LoaderManager.LoaderCallbacks<HashMap<String, Object>>() {
            @Override
            public Loader<HashMap<String, Object>> onCreateLoader(int id, Bundle args) {
                HTTPClient httpClient = new HTTPClient(activity);
                httpClient.getPredictions(args.getString("page"), args.getString("pair"));
                return httpClient;
            }

            @Override
            public void onLoadFinished(Loader<HashMap<String, Object>> loader, HashMap<String, Object> data) {
                int code = Integer.parseInt(data.get("statusCode").toString());
                if(code == 200) {
                    try {
                        JSONObject body = new JSONObject(data.get("body").toString());
                        JSONArray jsonArray = body.getJSONArray("predictions");
                        final ArrayList<Prediction> predictions = new ArrayList<>();
                        for(int i=0;i<jsonArray.length();i++) {
                            predictions.add(new Prediction(jsonArray.getJSONObject(i)));
                        }
                        if(!pair.equals(predictionView.getCurrentPair())) {
                            predictionView.clearListView();
                            predictionView.setCurrentPair(pair);
                        }
                        predictionView.addPredictions(predictions);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                } else {
                    predictionView.showMessage("予測情報の取得に失敗しました");
                }
                getLoaderManager().destroyLoader(LOADER_ID);
            }

            public void onLoaderReset(Loader<HashMap<String, Object>> loader) {}
        });
    }
}


app/src/main/java/com/leonis/android/rasalas/lib/HTTPClient.java
package com.leonis.android.rasalas.lib;

import android.content.AsyncTaskLoader;
import android.content.Context;
import android.util.Base64;

import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.HashMap;
import java.util.Properties;

/**
 * Created by leonis on 2019/09/22.
 */

public class HTTPClient extends AsyncTaskLoader<HashMap<String, Object>> {
    private static final String BASE_PATH = "/regulus/api";
    private static final String PORT = "80";

    private HttpURLConnection con;
    private HashMap<String, Object> response;
    private String baseUrl;
    private String credential;

    public HTTPClient(Context context) {
        super(context);
        InputStream inputStream = context.getClassLoader().getResourceAsStream("web-api.properties");

        try {
            Properties webApiProp = new Properties();
            webApiProp.load(inputStream);
            baseUrl = "http://" + webApiProp.getProperty("host") + ":" + PORT + BASE_PATH;

            String application_id = webApiProp.getProperty("application_id");
            String application_key = webApiProp.getProperty("application_key");
            byte[] credential = Base64.encode((application_id + ":" + application_key).getBytes(), Base64.DEFAULT);
            this.credential = new String(credential, "UTF-8");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void getPredictions(String page, String pair) {
        try {
            StringBuilder query_string = new StringBuilder("?means=auto&page=");
            query_string.append(page);
            query_string.append("&pair=");
            query_string.append(pair);
            con = (HttpURLConnection) new URL(baseUrl + "/predictions" + query_string).openConnection();
            con.setRequestMethod("GET");
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private HashMap<String, Object> sendRequest() {
        try {
            con.setRequestProperty("Authorization", "Basic " + credential);
            con.connect();

            StringBuilder sb = new StringBuilder();
            String st;
            BufferedReader br;
            try {
                br = new BufferedReader(new InputStreamReader(con.getInputStream(), "UTF-8"));
            } catch (FileNotFoundException e) {
                br = new BufferedReader(new InputStreamReader(con.getErrorStream(), "UTF-8"));
            }
            while((st = br.readLine()) != null){
                sb.append(st);
            }
            response = new HashMap<>();
            response.put("statusCode", con.getResponseCode());
            response.put("body", sb.toString());
        } catch (MalformedURLException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return response;
    }

    @Override
    public HashMap<String, Object> loadInBackground() {
        return sendRequest();
    }

    @Override
    protected void onStartLoading() {
        forceLoad();
    }
}


app/src/main/java/com/leonis/android/rasalas/models/Prediction.java
package com.leonis.android.rasalas.models;

import org.json.JSONException;
import org.json.JSONObject;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

/**
 * Created by leonis on 2019/09/22.
 */

public class Prediction {
    private Date datetime;
    private String pair;
    private String result;

    public Prediction(JSONObject prediction) {
        try {
            SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.JAPAN);
            this.datetime = formatter.parse(prediction.getString("created_at"));
            this.pair = prediction.getString("pair");
            this.result = prediction.getString("result");
        } catch (JSONException ignored) {
        } catch (ParseException ignored) {
        }
    }

    public Date getDatetime() {
        return this.datetime;
    }

    public String getPair() {
        return this.pair;
    }

    public String getResult() {
        return this.result;
    }
}


app/src/main/java/com/leonis/android/rasalas/views/PredictionView.java
package com.leonis.android.rasalas.views;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.Button;
import android.widget.ListAdapter;
import android.widget.ListView;
import android.widget.RelativeLayout;
import android.widget.Spinner;
import android.widget.Toast;

import com.leonis.android.rasalas.MainActivity;
import com.leonis.android.rasalas.R;
import com.leonis.android.rasalas.models.Prediction;

import java.util.ArrayList;

/**
 * Created by leonis on 2019/09/22.
 */

public class PredictionView extends RelativeLayout implements OnClickListener, OnItemSelectedListener {
    public static final String DEFAULT_PAGE = "1";
    public static final String DEFAULT_PAIR = "USDJPY";

    private final PredictionListAdapter predictionListAdapter;
    private final ListView predictionListView;
    private final ArrayList<Prediction> predictions;
    private final Spinner pairs;
    private final Button nextPage;
    private int currentPage;
    private String currentPair;

    private final Context context;

    public PredictionView(Context context, AttributeSet attributeSet) {
        super(context, attributeSet);
        this.context = context;

        View layout = View.inflate(context, R.layout.prediction_view, this);

        nextPage = layout.findViewById(R.id.next_page);
        nextPage.setOnClickListener(this);
        nextPage.setVisibility(INVISIBLE);

        pairs = layout.findViewById(R.id.prediction_select_pair);
        pairs.setOnItemSelectedListener(this);

        predictions = new ArrayList<>();
        predictionListAdapter = new PredictionListAdapter(context, predictions);
        predictionListView = layout.findViewById(R.id.prediction_list);

        currentPage = 1;
        currentPair = DEFAULT_PAIR;
    }

    public void setCurrentPair(String currentPair) {
        this.currentPair = currentPair;
    }

    public String getCurrentPair() {
        return currentPair;
    }

    public void showMessage(String message) {
        Toast.makeText(context, message, Toast.LENGTH_LONG).show();
    }

    public void addPredictions(ArrayList<Prediction> predictions) {
        this.predictions.addAll(predictions);
        predictionListView.setAdapter(predictionListAdapter);
        fixListViewHeight(predictionListView);
        nextPage.setVisibility(predictions.isEmpty() ? INVISIBLE : VISIBLE);
    }

    public void clearListView() {
        predictions.clear();
    }

    private void fixListViewHeight(ListView listView) {
        int totalHeight = 0;
        ListAdapter adapter = listView.getAdapter();
        int itemCount = adapter.getCount();

        for (int i = 0;i < itemCount;i++) {
            View listItem = adapter.getView(i, null, listView);
            listItem.measure(0, 0);
            totalHeight += listItem.getMeasuredHeight() + 3;
        }

        ViewGroup.LayoutParams params = listView.getLayoutParams();
        params.height = totalHeight;
        listView.setLayoutParams(params);
    }

    @Override
    public void onClick(View v) {
        ((MainActivity) context).getPredictions(Integer.toString(++currentPage), currentPair);
    }

    @Override
    public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
        Spinner spinner = (Spinner) adapterView;
        String selectedPair = spinner.getSelectedItem().toString();
        if(!currentPair.equals(selectedPair)) {
            ((MainActivity) context).getPredictions(DEFAULT_PAGE, selectedPair);
        }
    }

    @Override
    public void onNothingSelected(AdapterView<?> adapterView) {}
}


app/src/main/java/com/leonis/android/rasalas/views/PredictionListAdapter.java
package com.leonis.android.rasalas.views;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.TextView;

import com.leonis.android.rasalas.R;
import com.leonis.android.rasalas.models.Prediction;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Locale;

/**
 * Created by leonis on 2019/09/22.
 */

class PredictionListAdapter extends BaseAdapter {
    private final Context context;
    private final List<Prediction> predictions;

    private class ViewHolder {
        TextView dateView;
        TextView pairView;
        ImageView resultView;
    }

    public PredictionListAdapter(Context context, List<Prediction> predictions) {
        this.context = context;
        this.predictions = predictions;
    }

    @Override
    public int getCount() {
        return predictions.size();
    }

    @Override
    public Object getItem(int position) {
        return predictions.get(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        View view = convertView;
        ViewHolder holder;

        Prediction prediction = predictions.get(position);

        if (view == null) {
            LayoutInflater inflater =
                    (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
            view = inflater != null ? inflater.inflate(R.layout.prediction_list_view, parent, false) : null;

            TextView dateView = view.findViewById(R.id.prediction_date);
            TextView pairView = view.findViewById(R.id.prediction_pair);
            ImageView resultView = view.findViewById(R.id.prediction_result);

            holder = new ViewHolder();
            holder.dateView = dateView;
            holder.pairView = pairView;
            holder.resultView = resultView;
            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }

        SimpleDateFormat formatter = new SimpleDateFormat("yyyy/MM/dd HH:mm:ss", Locale.JAPAN);
        holder.dateView.setText(formatter.format(prediction.getDatetime()));
        holder.pairView.setText(prediction.getPair());
        switch (prediction.getResult()) {
            case "up":
                holder.resultView.setImageResource(R.mipmap.up);
                break;
            case "down":
                holder.resultView.setImageResource(R.mipmap.down);
                break;
            default :
                holder.resultView.setImageResource(R.mipmap.range);
        }

        return view;
    }
}


app/src/androidTest/java/com/leonis/android/rasalas/lib/HTTPClientTest.java
package com.leonis.android.rasalas.lib;

import android.support.test.runner.AndroidJUnit4;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Field;
import java.net.HttpURLConnection;
import java.util.HashMap;
import java.util.Map;

import static android.support.test.InstrumentationRegistry.getContext;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * Created by leonis on 2019/09/24.
 */

@RunWith(AndroidJUnit4.class)
public class HTTPClientTest {
    private HashMap<String, Object> ret;
    private HttpURLConnection con;
    private JSONObject responseBody;

    public HTTPClientTest() {}

    @Before
    public void setUp() throws Exception {
        con = mock(HttpURLConnection.class);
        doNothing().when(con).connect();
        responseBody = new JSONObject();
    }

    @Test
    public void testGetPredictions_OK() {
        HTTPClient httpClient = new HTTPClient(getContext());
        httpClient.getPredictions("1", "USDJPY");
        JSONArray predictions = new JSONArray();
        JSONObject prediction = new JSONObject();
        HashMap<String, String> expected = new HashMap<String, String>() {
            {
                put("prediction_id", "60a830b4b980f68ab0828517f806e680");
                put("model", "model.zip");
                put("from", "2019-08-01T00:00:00.000Z");
                put("to", "2019-08-01T23:59:59.000Z");
                put("pair", "USDJPY");
                put("means", "auto");
                put("result", "up");
                put("state", "completed");
                put("created", "2019-08-02T09:00:00.000Z");
            }
        };

        try {
            for (Map.Entry<String, String> entry : expected.entrySet()) {
                prediction.put(entry.getKey(), entry.getValue());
            }
            predictions.put(prediction);
            responseBody.put("predictions", predictions);
        } catch (JSONException e) {
            e.printStackTrace();
            fail();
        }
        setupMock(httpClient, 200, responseBody.toString());

        ret = httpClient.loadInBackground();

        assertEquals(200, Integer.parseInt(ret.get("statusCode").toString()));

        try {
            responseBody = new JSONObject(ret.get("body").toString());
            predictions = responseBody.getJSONArray("predictions");
            prediction = predictions.getJSONObject(0);
            for (Map.Entry<String, String> entry : expected.entrySet()) {
                assertEquals(entry.getValue(), prediction.getString(entry.getKey()));
            }
        } catch (JSONException e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testGetPredictions_NG_invalidPage() {
        HTTPClient httpClient = new HTTPClient(getContext());
        httpClient.getPredictions("invalid", "USDJPY");
        JSONArray errors = new JSONArray();
        JSONObject error = new JSONObject();

        try {
            error.put("error_code", "invalid_params_page");
            errors.put(error);
            responseBody.put("errors", errors);
        } catch (JSONException e) {
            e.printStackTrace();
            fail();
        }
        setupMock(httpClient, 400, responseBody.toString());

        ret = httpClient.loadInBackground();

        assertEquals(400, Integer.parseInt(ret.get("statusCode").toString()));

        try {
            responseBody = new JSONObject(ret.get("body").toString());
            errors = responseBody.getJSONArray("errors");
            error = errors.getJSONObject(0);

            assertEquals("invalid_params_page", error.getString("error_code"));
        } catch (JSONException e) {
            e.printStackTrace();
            fail();
        }
    }

    @Test
    public void testGetPredictions_NG_invalidPair() {
        HTTPClient httpClient = new HTTPClient(getContext());
        httpClient.getPredictions("1", "invalid");
        JSONArray errors = new JSONArray();
        JSONObject error = new JSONObject();

        try {
            error.put("error_code", "invalid_params_pair");
            errors.put(error);
            responseBody.put("errors", errors);
        } catch (JSONException e) {
            e.printStackTrace();
            fail();
        }
        setupMock(httpClient, 400, responseBody.toString());

        ret = httpClient.loadInBackground();

        assertEquals(400, Integer.parseInt(ret.get("statusCode").toString()));

        try {
            responseBody = new JSONObject(ret.get("body").toString());
            errors = responseBody.getJSONArray("errors");
            error = errors.getJSONObject(0);

            assertEquals("invalid_params_pair", error.getString("error_code"));
        } catch (JSONException e) {
            e.printStackTrace();
            fail();
        }
    }

    private void setupMock(HTTPClient httpClient, int statusCode, final String responseBody) {
        try {
            when(con.getRequestMethod()).thenReturn("");
            when(con.getResponseCode()).thenReturn(statusCode);
            when(con.getInputStream()).thenReturn(new InputStream() {
                private int position = 0;
                @Override
                public int read() throws IOException {
                    return position < responseBody.length() ? responseBody.charAt(position++) : -1;
                }

                @Override
                public void close() throws IOException {}
            });

            Class<? extends HTTPClient> c = httpClient.getClass();
            Field f = c.getDeclaredField("con");
            f.setAccessible(true);
            f.set(httpClient, con);
        } catch (NoSuchFieldException e) {
            e.printStackTrace();
        } catch (IllegalArgumentException e) {
            e.printStackTrace();
        } catch (IllegalAccessException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
